Automattic/wp-calypso
# composite-checkout: Add success/info/error messaging

#### Changes proposed in this Pull Request

The Checkout form currently has the `onSuccess` and `onFailure` callbacks. However, these actually serve quite different purposes. When the payment is successful, we'll call the `onSuccess` callback to redirect to a post-purchase page or perform some other operation. When there is a failure, we keep the form visible, and pass an error object to `onFailure` to notify the user. In addition, there may be cases where we want to notify the user of some other intermediate success or informational message (eg: "your coupon was successfully added" or "the item was removed from the cart") for which we cannot use either of the existing callbacks.

To resolve all these issues, this PR renames `onSuccess` to `onPaymentComplete`, renames `onFailure` to `showErrorMessage`, and adds two additional (required) callbacks, `showInfoMessage` and `showSuccessMessage`.

The original two callbacks (as well as the arbitrary event function `onEvent`) were previously accessed with the `useCheckoutHandlers` hook. This PR replaces that hook with `usePaymentComplete` for `onPaymentComplete`, `useEvents` for `onEvent`, and `useMessages` for the other three.

#### Testing instructions

- Run the demo using `npm run composite-checkout-demo`.
- Load the checkout form demo using the second credit card method. Fill out the form but set the country to "-". Press "Pay". Verify that you see an error message in the console beginning with the word "Error". This validates `showErrorMessage`.
- Modify `packages/composite-checkout/src/lib/payment-methods/stripe-credit-card-fields.js` to replace `showErrorMessage` with `showInfoMessage`. Repeat the above step but verify that the error message begins with the word "Info". This validates `showInfoMessage`.
- Repeat the last step but replace the function with a call to `showSuccessMessage` and verify that the error message begins with the word "Success". This validates `showSuccessMessage`.
- Complete one step and press "Continue". Verify that the console shows a message that the step changed. This validates `onEvent`.
- Complete a purchase using the credit card form. Verify that a success message is shown. This validates `onPaymentComplete`.
